bliss/models/galsim_decoder.py

```diff
@@ -0,0 +1,291 @@
+            disk = galsim.Exponential(flux=disk_flux, half_light_radius=disk_hlr_arcsecs).shear(
+                q=disk_q,
+                beta=beta_radians * galsim.radians,
+            )
+            components.append(disk)
+        if bulge_flux > 0:
+            b_b = bulge_q * a_b
+            bulge_hlr_arcsecs = np.sqrt(a_b * b_b)
+            bulge = galsim.DeVaucouleurs(
+                flux=bulge_flux, half_light_radius=bulge_hlr_arcsecs
+            ).shear(q=bulge_q, beta=beta_radians * galsim.radians)
+            components.append(bulge)
+        galaxy = galsim.Add(components)
+        gal_conv = galsim.Convolution(galaxy, psf)
+        offset = offset if offset is None else offset.numpy()
+        image = gal_conv.drawImage(
+            nx=slen, ny=slen, method="auto", scale=self.pixel_scale, offset=offset
+        )
+        return torch.from_numpy(image.array).reshape(1, slen, slen)
+
+
+class UniformGalsimGalaxiesPrior:
+    def __init__(
+        self,
+        single_galaxy_prior: SingleGalsimGalaxyPrior,
+        max_n_sources: int,
+        max_shift: float,
+    ):
+        self.single_galaxy_prior = single_galaxy_prior
+        self.max_shift = max_shift  # between 0 and 0.5, from center.
+        self.max_n_sources = max_n_sources
+        self.dim_latents = self.single_galaxy_prior.dim_latents
+        assert 0 <= self.max_shift <= 0.5
+
+    def sample(self) -> Dict[str, Tensor]:
+        """Returns a single batch of source parameters."""
+        n_sources = _sample_n_sources(self.max_n_sources)
+
+        params = torch.zeros(self.max_n_sources, self.dim_latents)
+        params[:n_sources, :] = self.single_galaxy_prior.sample(n_sources)
+
+        locs = torch.zeros(self.max_n_sources, 2)
+        locs[:n_sources, 0] = _uniform(-self.max_shift, self.max_shift, n_sources) + 0.5
+        locs[:n_sources, 1] = _uniform(-self.max_shift, self.max_shift, n_sources) + 0.5
+
+        # for now, galaxies only
+        galaxy_bools = torch.zeros(self.max_n_sources, 1)
+        galaxy_bools[:n_sources, :] = 1
+        star_bools = torch.zeros(self.max_n_sources, 1)
+
+        return {
+            "n_sources": torch.tensor(n_sources),
+            "galaxy_params": params,
+            "locs": locs,
+            "galaxy_bools": galaxy_bools,
+            "star_bools": star_bools,
+        }
+
+
+class UniformBrightestCenterGalsimGalaxy(UniformGalsimGalaxiesPrior):
+    def sample(self) -> Dict[str, Tensor]:
+        """Returns a single batch of source parameters where brightest galaxy is centered."""
+        sample = super().sample()
+        n_sources = sample.pop("n_sources")
+        flux = sample["galaxy_params"][:, 0].reshape(-1)
+        idx_order = torch.argsort(-flux)
+        reordered_sample = {k: v[idx_order] for k, v in sample.items()}
+        reordered_sample["locs"][0, :] = 0.5  # centered
+        return {"n_sources": n_sources, **reordered_sample}
+
+
+class GalsimGalaxiesDecoder:
```

Comment:
I assume this is the new decoder (i.e. meant to replace the "ImageDecoder")

Reply:
That's right!

Reply:
and here is the associated [dataset](https://github.com/prob-ml/bliss/pull/640/files#diff-1a8faf476915ed024d8eedd8cbd2d8c94bfc57ba4b79c57280776194793d54f1R77) that combines the prior and the decoder